src/main/scala/com.snowplowanalytics.sauna/responders/urbanairship/UAResponder.scala

```diff
@@ -44,57 +46,72 @@ class UAResponder(urbanairship: urbanAirship)(implicit logger: ActorRef) extends
-      } else {
-
-        if (listToIdentifierMap.contains(listName)) {
-          listToIdentifierMap = appToListMap(appKey)
-
-          listToIdentifierMap += (listName -> (listToIdentifierMap(listName) :+ new Airship(idType, id)))
-        } else {
-
-          listToIdentifierMap += (listName -> List(new Airship(idType, id)))
-        }
-      }
-
-      appToListMap += (appKey -> listToIdentifierMap)
-      count += 1
-
-      if (count >= 10000000) {
-        urbanairship.maptoRequest(appToListMap)
-        appToListMap = Map()
-        count = 0
+    for(rawLine <- fromInputStream(is).getLines)
+    {
+      val (listToIdentifierMap,appKey) = getMapEntryForLine(rawLine,appToListMap)  
+     if(appKey!=null)
+     {
+       appToListMap += (appKey -> listToIdentifierMap)
+        count += 1
+        if (count >= 10000000) {
+          urbanairship.maptoRequest(appToListMap)
+          appToListMap = Map()
+          count = 0
+         }
       }
     }
-
     if (appToListMap.size > 0)
       urbanairship.maptoRequest(appToListMap)
-
   }
+  
+  def getMapEntryForLine(rawLine:String,appToListMap:Map[String, Map[String, List[Airship]]]):(Map[String, List[Airship]],String) = {
+      val uaInputOption:Option[UAInput] =  getEntryFromLine(rawLine)
+      if(!uaInputOption.isDefined)
+      {
+        logger ! Notification("not enough fields !")
+        return (null,null)
+      }
+      val uaInput:UAInput = uaInputOption.get
+      val listToIdentifierMap:Map[String, List[Airship]] = if (!appToListMap.contains(uaInput.appKey.get)) {
+         Map[String, List[Airship]](uaInput.listName.get -> List(new Airship(uaInput.idType.get, uaInput.id.get)))
+      } else {
+        val listMapForApp:Map[String, List[Airship]] = appToListMap(uaInput.appKey.get)
+        val listMap:Map[String, List[Airship]] = if (listMapForApp.contains(uaInput.listName.get)) {
+        val newListEntry:Map[String, List[Airship]]= Map(uaInput.listName.get -> (listMapForApp(uaInput.listName.get) :+ new Airship(uaInput.idType.get, uaInput.id.get)))  
+           listMapForApp ++ newListEntry
+        } else {
+          val newListEntry:Map[String, List[Airship]]= Map(uaInput.listName.get -> (List[Airship]() :+ new Airship(uaInput.idType.get, uaInput.id.get)))
+            listMapForApp ++ newListEntry 
+        }
+        listMap
+      }    
+      (listToIdentifierMap,uaInput.appKey.get)
+  }
+  
+  def getEntryFromLine(rawLine:String):Option[UAInput]={
+    
+    val line = rawLine.replace("\"", "").trim
+      
+     line.split("\t", -1) match 
+     { 
+        case Array(appKey, listName,idType , id) => ( Some(UAInput(Some(appKey),Some(listName),Some(idType),Some(id))))
+        case _ => (None) 
+     }
+  }
+  
 
 }
 
 object UAResponder {
 
   case class Airship(identifierType: String, identifier: String)
 
+  case class UAInput(appKey:Option[String],listName:Option[String],idType: Option[String], id: Option[String])
```

Comment:
Ok, @rmanojcit. One problem with this class is `Option`. `Option` denote that value can be absent. For example, if `getEntryFromLine` accepts line with 3 columns it returns `None`, which means `UAInput` wasn't found, and not `UAInput(Some(???), Some(???), Some(???), None))`. Because input line with 3 columns is actually absent (invalid) input, we cannot do anything with it.

This also eliminates need of `Option.get` above. Because we never use `.get` on `Option`, otherwise `Option` makes no sense, because they help us to avoid runtime-exceptions and runtime exception is exactly what gonna happen if you run `None.get`. So, instead of `Option.get` we use following approaches:

* Pattern matching. If you have good default action for it.
```scala
(maybeValue: Option[SomeType]) match {
  case Some(value) => process(value)
  case None => processDefault
}
```
* `.map`. We can pretend that we have a value.
```scala
(maybeValue: Option[SomeType]).map { (value: SomeType) =>
  // here we working with value, like we really have it
  // and if actually we don't have it - this function just won't get executed
  process(value)
}
```
```scala
// Our function can be even defined without any knowledge about Option
def process(value: SomeType) = ???
maybeValue.map(process)  // this is an essence of functional composition
```

* `.flatMap` and `for`-comprehensions. This coolest thing ever in Scala and functional languages. But for now pattern matching and `map` should be enough.

You can read more about `Option`s here: http://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html

One more thing is styling. Remember, we preserve same amount of spaces between similar literals? `appKey:Option[String],<space>listName:Option[String],<space>idType: Option[String]...`. And we write documentation for classes and functions even if their purpose is obvious.